File: compiler/syntax/parser.go
package syntax

import (
	"fmt"
	"reflect"

	db "wipple/database"
)

type ParseFunc[T any] func(*Parser) (T, *Error)

type Parser struct {
	Db     *db.Db
	Path   string
	Source string
	tokens []*Token
	index  int
	stack  []*commitEntry
	cache  map[uintptr]map[int]parseResult
}

type commitEntry struct {
	trace string
}

type parseResult struct {
	index int
	value any
}

func NewParser(db *db.Db, path string, source string) (*Parser, *Error) {
	parser := &Parser{
		Db:     db,
		Path:   path,
		Source: source,
		cache:  map[uintptr]map[int]parseResult{},
	}

	var err *Error
	parser.tokens, err = Tokenize(path, source)
	if err != nil {
		return nil, err
	}

	return parser, nil
}

func (parser *Parser) Spanned() func() db.Span {
	index := parser.index

	return func() db.Span {
		start := parser.eofSpan()
		if index < len(parser.tokens) {
			start = parser.tokens[index].span
		}

		end := parser.eofSpan()
		if parser.index > 0 {
			end = parser.tokens[parser.index-1].span
		}

		return db.JoinSpans(start, end, parser.Source)
	}
}

func (parser *Parser) backtrack(index int) {
	parser.index = index
}

func (parser *Parser) Error(message string) *Error {
	return parser.ErrorWithReason(message, "")
}

func (parser *Parser) ErrorWithReason(message string, reason string) *Error {
	span := parser.eofSpan()
	if parser.index < len(parser.tokens) {
		span = parser.tokens[parser.index].span
	}

	return &Error{
		Message: message,
		Reason:  reason,
		Span:    span,
	}
}

type TokenConfig struct {
	Name   string
	Reason string
}

func (parser *Parser) Token(kind string, configs ...TokenConfig) (string, *Error) {
	var config TokenConfig
	if len(configs) > 0 {
		config = configs[0]
	}

	expected := config.Name
	if expected == "" {
		expected = tokenNames[kind]
	}

	reason := config.Reason
	if reason == "" {
		reason = defaultTokenReasons[kind]
	}

	if parser.index >= len(parser.tokens) {
		return "", &Error{
			Message: fmt.Sprintf("Expected %s", kind),
			Reason:  config.Reason,
			Span:    parser.eofSpan(),
		}
	}

	token := parser.tokens[parser.index]

	if token.kind != kind {
		return "", &Error{
			Message: fmt.Sprintf("Expected %s, but found %s", expected, tokenNames[token.kind]),
			Reason:  reason,
			Span:    token.span,
		}
	}

	parser.index += 1

	return token.value, nil
}

func (parser *Parser) Commit(trace string) {
	parser.stack[len(parser.stack)-1].trace = trace
}

func (parser *Parser) ConsumeLineBreaks() {
	parser.Token("LineBreak")
}

func (parser *Parser) eofSpan() db.Span {
	if len(parser.tokens) == 0 {
		return db.NullSpan()
	} else {
		return parser.tokens[len(parser.tokens)-1].span
	}
}

func (parser *Parser) Finish() *Error {
	if parser.index < len(parser.tokens) {
		token := parser.tokens[parser.index]
		return parser.Error(fmt.Sprintf("Unexpected %s", tokenNames[token.kind]))
	}

	return nil
}

func ParseNothing(parser *Parser) (struct{}, *Error) {
	return struct{}{}, nil
}

func ParseCached[T any](parser *Parser, f ParseFunc[T]) (T, *Error) {
	start := parser.index

	key := reflect.ValueOf(f).Pointer()
	entry, ok := parser.cache[key]
	if !ok {
		entry = map[int]parseResult{}
		parser.cache[key] = entry
	}

	cached, ok := entry[start]
	if ok {
		parser.index = cached.index
		return cached.value.(T), nil
	}

	result, err := f(parser)
	if err != nil {
		return result, err
	}

	entry[start] = parseResult{
		index: parser.index,
		value: result,
	}

	return result, nil
}

func ParseOptional[T any](parser *Parser, f ParseFunc[T]) (T, bool, *Error) {
	start := parser.index

	entry := &commitEntry{}
	parser.stack = append(parser.stack, entry)
	defer func() {
		parser.stack = parser.stack[:len(parser.stack)-1]
	}()

	var result T
	result, err := f(parser)
	if err != nil {
		if entry.trace != "" {
			err.Committed = entry.trace
		}

		if err.Committed != "" {
			return result, false, err
		}

		parser.backtrack(start)
		return result, false, nil
	}

	return result, true, nil
}

type Many[T any, S any] struct {
	Value     T
	Separator S
}

func ParseMany[T any, S any](parser *Parser, min int, f ParseFunc[T], separator ParseFunc[S]) ([]Many[T, S], *Error) {
	first := true

	var results []Many[T, S]
	for {
		start := parser.index

		var separatorResult S
		if !first {
			var ok bool
			var err *Error
			separatorResult, ok, err = ParseOptional(parser, separator)
			if err != nil {
				return nil, err
			}

			if !ok {
				break
			}
		}

		result, ok, err := ParseOptional(parser, f)
		if err != nil {
			return nil, err
		}

		if !ok {
			parser.backtrack(start)
			break
		}

		results = append(results, Many[T, S]{
			Value:     result,
			Separator: separatorResult,
		})

		first = false
	}

	if len(results) < min {
		return nil, &Error{
			Message: fmt.Sprintf("Expected at least %d items", min),
		}
	}

	return results, nil
}

func ParseLines[T any](parser *Parser, min int, require bool, f ParseFunc[T]) ([]T, *Error) {
	parser.ConsumeLineBreaks()

	result, err := ParseMany(parser, min, f, func(parser *Parser) (*struct{}, *Error) {
		if require {
			_, err := parser.Token("LineBreak")
			if err != nil {
				return nil, err
			}
		} else {
			parser.ConsumeLineBreaks()
		}

		return &struct{}{}, nil
	})

	if err != nil {
		return nil, err
	}

	parser.ConsumeLineBreaks()

	values := make([]T, 0, len(result))
	for _, item := range result {
		values = append(values, item.Value)
	}

	return values, nil
}

File: compiler/src/syntax/parser.rs
use crate::{
    database::{Db, Node, NodeRef, Span},
    syntax::{ParseError, Token, TokenKind, tokenize},
};
use std::{
    any::{Any, TypeId},
    collections::HashMap,
};

#[derive(Debug)]
pub struct Parser<'db> {
    db: &'db mut Db,
    source: String,
    tokens: Vec<Token>,
    index: usize,
    stack: Vec<StackEntry>,
    cache: Cache,
}

#[derive(Debug, Default)]
struct StackEntry {
    trace: Option<String>,
}

#[derive(Debug)]
struct CacheEntry {
    index: usize,
    value: Box<dyn Any>,
}

#[derive(Debug, Default)]
struct Cache(HashMap<TypeId, HashMap<usize, CacheEntry>>);

impl<'db> Parser<'db> {
    pub fn new(db: &'db mut Db, path: &str, source: &str) -> Result<Self, ParseError> {
        let tokens = tokenize(path, source)?;

        Ok(Self {
            db,
            source: source.to_string(),
            tokens,
            index: 0,
            stack: Vec::new(),
            cache: Cache(HashMap::new()),
        })
    }

    pub fn spanned(&self) -> impl FnOnce(&mut Self) -> Span + use<'db> {
        let start = self
            .tokens
            .get(self.index)
            .map_or_else(|| self.eof_span(), |token| token.span.clone());

        move |parser| {
            let end = parser
                .tokens
                .get(parser.index.saturating_sub(1))
                .map_or_else(|| parser.eof_span(), |token| token.span.clone());

            Span::join(start, end, &parser.source)
        }
    }

    pub fn backtrack(&mut self, index: usize) {
        self.index = index;
    }

    pub fn error(&self, message: impl Into<String>) -> ParseError {
        self.error_with_reason(message, None)
    }

    pub fn error_with_reason(
        &self,
        message: impl Into<String>,
        reason: Option<String>,
    ) -> ParseError {
        let span = self
            .tokens
            .get(self.index)
            .map_or_else(|| self.eof_span(), |token| token.span.clone());

        ParseError {
            message: message.into(),
            reason,
            committed: None,
            span,
        }
    }

    pub fn token(&mut self, kind: TokenKind) -> Result<String, ParseError> {
        self.token_with(kind, None, None)
    }

    pub fn token_with_name(&mut self, kind: TokenKind, name: &str) -> Result<String, ParseError> {
        self.token_with(kind, Some(name), None)
    }

    pub fn token_with_reason(
        &mut self,
        kind: TokenKind,
        reason: &str,
    ) -> Result<String, ParseError> {
        self.token_with(kind, None, Some(reason))
    }

    pub fn token_with(
        &mut self,
        kind: TokenKind,
        name: Option<&str>,
        reason: Option<&str>,
    ) -> Result<String, ParseError> {
        let expected = name.unwrap_or_else(|| kind.name());

        let Some(token) = &self.tokens.get(self.index) else {
            return Err(ParseError {
                message: format!("Expected {}", expected),
                reason: reason.map(|s| s.to_string()),
                committed: None,
                span: self.eof_span(),
            });
        };

        if token.kind != kind {
            return Err(ParseError {
                message: format!("Expected {}, but found {}", expected, token.kind.name()),
                reason: reason.map(|s| s.to_string()),
                committed: None,
                span: token.span.clone(),
            });
        }

        self.index += 1;

        Ok(token.value.clone())
    }

    pub fn try_node<T: Node>(
        &mut self,
        f: impl FnOnce(&mut Parser<'_>) -> Result<T, ParseError>,
    ) -> Result<Option<NodeRef>, ParseError> {
        self.parse_optional(|parser| parser.node(f))
    }

    pub fn node<T: Node>(
        &mut self,
        f: impl FnOnce(&mut Parser<'_>) -> Result<T, ParseError>,
    ) -> Result<NodeRef, ParseError> {
        let span = self.spanned();
        let node = f(self)?;
        let span = span(self);
        Ok(self.register(span, node))
    }

    pub fn register<T: Node>(&mut self, span: Span, node: T) -> NodeRef {
        self.db.node(span, node)
    }

    pub fn span(&self, node: &NodeRef) -> Span {
        self.db.span(node)
    }

    pub fn join_spans(&self, left: &Span, right: &Span) -> Span {
        Span::join(left.clone(), right.clone(), &self.source)
    }

    pub fn commit(&mut self, trace: &str) {
        self.stack.last_mut().unwrap().trace = Some(trace.to_string());
    }

    pub fn consume_line_breaks(&mut self) {
        let _ = self.token(TokenKind::LineBreak);
    }

    fn eof_span(&self) -> Span {
        self.tokens
            .last()
            .map_or_else(Span::empty, |token| token.span.clone())
    }

    pub fn finish(self) -> Result<(), ParseError> {
        if let Some(token) = self.tokens.get(self.index) {
            return Err(self.error(format!("Unexpected {}", token.kind.name())));
        }

        Ok(())
    }

    pub fn parse_nothing(&mut self) -> Result<(), ParseError> {
        Ok(())
    }

    pub fn parse_cached<
        T: Clone + 'static,
        F: Fn(&mut Parser<'_>) -> Result<T, ParseError> + 'static,
    >(
        &mut self,
        f: F,
    ) -> Result<T, ParseError> {
        let key = TypeId::of::<F>();
        let start = self.index;

        if let Some(cached) = self.cache.0.entry(key).or_default().get(&start) {
            let value = cached.value.downcast_ref::<T>().unwrap();
            self.index = cached.index;
            return Ok(value.clone());
        }

        let result = f(self)?;

        self.cache.0.entry(key).or_default().insert(
            start,
            CacheEntry {
                index: self.index,
                value: Box::new(result.clone()),
            },
        );

        Ok(result)
    }

    pub fn parse_optional<T>(
        &mut self,
        f: impl FnOnce(&mut Parser<'_>) -> Result<T, ParseError>,
    ) -> Result<Option<T>, ParseError> {
        let start = self.index;

        let entry = StackEntry::default();
        self.stack.push(entry);
        let result = f(self);
        let entry = self.stack.pop().unwrap();

        match result {
            Ok(value) => Ok(Some(value)),
            Err(mut error) => {
                if let Some(trace) = entry.trace {
                    error.committed = Some(trace);
                }

                if error.committed.is_some() {
                    return Err(error);
                }

                self.backtrack(start);

                Ok(None)
            }
        }
    }

    pub fn parse_many<T, S: Default>(
        &mut self,
        min: usize,
        f: impl Fn(&mut Parser<'_>) -> Result<T, ParseError>,
        separator: impl Fn(&mut Parser<'_>) -> Result<S, ParseError>,
    ) -> Result<Vec<(T, (S, Span))>, ParseError> {
        let mut first = true;

        let mut results = Vec::new();
        loop {
            let start = self.index;

            let span = self.spanned();

            let mut separator_result = S::default();
            if !first {
                match self.parse_optional(&separator)? {
                    Some(result) => separator_result = result,
                    None => break,
                }
            }

            let separator_span = span(self);

            let result = match self.parse_optional(&f)? {
                Some(result) => result,
                None => {
                    self.backtrack(start);
                    break;
                }
            };

            results.push((result, (separator_result, separator_span)));

            first = false;
        }

        if results.len() < min {
            let span = match results.last() {
                Some((_, (_, span))) => span.clone(),
                None => self.eof_span(),
            };

            return Err(ParseError {
                message: format!("Expected at least {} items", min),
                reason: None,
                committed: None,
                span,
            });
        }

        Ok(results)
    }

    pub fn parse_lines<T, F>(
        &mut self,
        min: usize,
        require: bool,
        f: F,
    ) -> Result<Vec<T>, ParseError>
    where
        F: Fn(&mut Parser<'_>) -> Result<T, ParseError>,
    {
        self.consume_line_breaks();

        let result = self.parse_many(min, f, |parser| {
            if require {
                parser.token(TokenKind::LineBreak)?;
            } else {
                parser.consume_line_breaks();
            }

            Ok(())
        })?;

        self.consume_line_breaks();

        Ok(result.into_iter().map(|(result, _)| result).collect())
    }
}
